File: clojure/shortest_path.clj
;;  b - d
;; /    |
;;a     e
;; \  /
;;  c 
(def graph {
    :a [:b :c],
    :b [:d],
    :c [:e],
    :d [:e],
    :e []
})

(declare shortest_path)

(defn shortest [a b]
    (if (< (count a) (count b))
        a
        b
    )
)

(defn shortest_non_empty [a b]
    (if (empty? a)
        b
        (if (empty? b)
            a 
            (shortest a b)
        )
    )
)

(defn shortest_path_aux [nodes node2]
    (if (empty? nodes)
        []
        (shortest_non_empty 
            (shortest_path (first nodes) node2)
            (shortest_path_aux (rest nodes) node2)
        )
    )
)

(defn shortest_path [node1 node2]
    (if (= node1 node2)
        [node2]
        (if (empty? (graph node1))
            []
            (concat
                [node1]
                (shortest_path_aux (graph node1) node2)
            )
        )
    )
)

(print (shortest_path :a :e))

File: python/shortest_path.py
#  b - d
# /    |
#a     e
# \  /
#  c 

# The graph spec
graph = {
    'a': ['b', 'c'],
    'b': ['d'],
    'c': [],
    'd': ['e'],
    'e': []
}

# Compute the shortest path recursively
def shortest_path(node1, node2):
    if (node1 == node2):
        return [node2]
    
    min_len = None
    shortest = []
    for node in graph[node1]:
        path = shortest_path(node, node2)
        if not path: continue

        path_length = len(path)
        # Initialize min_length
        if not min_len:
            min_len = path_length
            shortest = path
        # New minimum
        elif path_length < min_len:
            min_len = path_length
            shortest = path

    # If there's a valid path, return the current node concatenated with the shortest path
    if shortest:
        return [node1] + shortest

print(shortest_path('a', 'e'))
